### PR TITLE
Add setting to have the volume of NVDA sounds follow the voice volume.

### LINKED ISSUE #1409: Ability to adjust volume of sounds
**Reported by KevanGC on 2011-03-09 06:12**
Hi,

I've noticed that when you're browsing the web and you switch between Browse and Focus mode, NVDA plays these sound-effects. The sound-effects are very loud and sometimes they get annoying. It would be cool if you could add a feature for users to adjust the volumes of sound effects and progress-bar tones?
**Blocking #892, #2978, #3100, #4590**

### PR DESCRIPTION
**Pull request by jcsteh**
### Link to issue number:
Fixes #1409.

### Summary of the issue:
Users would like to be able to control the volume of NVDA sounds. With the WASAPI implementation in #14697, it is possible to separately control the volume of NVDA sounds using the Windows Volume Mixer. However, it is often useful to link the volume of NVDA sounds to the volume of the voice.

### Description of user facing changes
It is now possible to have the volume of NVDA sounds and beeps follow the volume setting of the voice you are using. This option can be enabled in Advanced settings.

### Description of development approach
When this option is enabled and a WasapiWavePlayer is using soundsSession, the open() and stop() methods of WasapiWavePlayer set the session volume to the synth's voice volume. Note that we do this in stop() because some players keep the device open; e.g. tones.

### Testing strategy:
1. Enabled the new setting.
2. Decreased the voice volume.
3. Switched between browse and focus modes. Observed that the sound volume was lower.
4. Moved the mouse with mouse beeps enabled. Observed that the beep volume was lower.
5. Increased the voice volume to maximum.
6. Switched between browse and focus modes. Observed that the sound volume was maximum.
7. Moved the mouse with mouse beeps enabled. Observed that the beep volume was maximum.
8. Disabled the new setting.
9. Decreased the voice volume.
10. Switched between browse and focus modes. Observed that the sound volume was still maximum.
11. Moved the mouse with mouse beeps enabled. Observed that the beep volume was still maximum.
12. Enabled the new setting.
13. Switched between browse and focus modes. Observed that the sound volume was lower.
14. Disabled the new setting.
15. Switched between browse and focus modes. Observed that the sound volume was maximum.
16. Enabled the new setting.
17. Disconnected the default audio device (a headset).
18. Moved the mouse with mouse beeps enabled. Observed that the beep volume was lower.

### Known issues with pull request:
1. This is in Advanced settings because it depends on WASAPI and WASAPI is in Advanced settings. Once WASAPI is the only option, we can move this to Voice settings. Or should we just move it now?
2. The volume of a sound isn't adjusted while it is playing if the voice volume is changed. NVDA sounds usually aren't (and shouldn't be) very long, so I don't think this is a real issue, but I thought it worth noting anyway.

### Change log entries:
New features

> It is now possible to have the volume of NVDA sounds and beeps follow the volume setting of the voice you are using. This option can be enabled in Advanced settings. (#1409)

Note that I forgot to mention that it's possible to separately control the volume of NVDA sounds in the What's New for #14697. We could add that too:

> You can now separately control the volume of NVDA sounds. This can be done using the Windows Volume Mixer. (#1409)

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

**Comment by XLTechie** [comment]
One other test case that should be here, I think: decrease the volume with the 
setting enabled, and confirm sounds are low. Then disable the setting and 
confirm that the voice is still low, but the sounds are high.

Personally, I think we need an audio settings panel, containing this setting, 
the WASAPI or not setting, and probably (eventually?) the sound output device 
setting that is currently found in the synth selection dialog.

There are other options that should probably end up there if eventually 
implemented, such as the ability to change certain tone pitches, that negatively 
impact people with hearing difficulties at their current fixed defaults.
(#10116, #9206)

File: source/gui/settingsDialogs.py [hunk 1]
```diff
@@ -3177,7 +3191,7 @@ def onSave(self):
 		config.conf["documentFormatting"]["reportTransparentColor"] = (
 			self.reportTransparentColorCheckBox.IsChecked()
 		)
-		config.conf["audio"]["wasapi"] = self.wasapiCheckBox.IsChecked()
+		config.conf["audio"]["soundVolumeFollowsVoice"] = self.soundVolFollowCheckBox.IsChecked()
```
Comment (seanbudd) [inline source/gui/settingsDialogs.py:3194]: it seems you accidentally removed the saving of the `wasapiCheckBox` value

File: user_docs/en/userGuide.t2t [hunk 2]
```diff
@@ -2299,6 +2299,13 @@ WASAPI is a more modern audio framework which may improve the responsiveness, pe
 This option is enabled by default.
 After changing this option, you will need to restart NVDA for the change to take effect.
 
+==== Volume of NVDA sounds follows voice volume ====[SoundVolumeFollowsVoice]
+When this option is enabled, the volume of NVDA sounds and beeps will follow the volume setting of the voice you are using.
+If you decrease the volume of the voice, the volume of sounds will decrease.
+Similarly, if you increase the volume of the voice, the volume of sounds will increase.
+This option only takes effect when Use WASAPI for audio output is enabled.
```
Comment (seanbudd) [inline user_docs/en/userGuide.t2t:2306]: Quotation marks here would improve readability

```suggestion
This option only takes effect when "Use WASAPI for audio output" is enabled.
```

**Comment by jcsteh** [comment]
> One other test case that should be here, I think: decrease the volume with the
setting enabled, and confirm sounds are low. Then disable the setting and
confirm that the voice is still low, but the sounds are high.

I intentionally didn't include this test case, as I know it won't work. Whether it should is an interesting question. Strictly speaking, disabling the setting should prevent the sound volume from following the voice volume, but it doesn't say anything about snapping the volume back to maximum. The user might have had the sound volume set to something lower before they enabled follow. However, it also doesn't really make sense to me to keep track of the level the user set before enabling follow just so we can restore it later.

**Comment by jcsteh** [comment]
If we did want to snap back to maximum, we'd need to do that when applying the disable change somehow.

**Comment by XLTechie** [comment]
> I intentionally didn't include this test case, as I know it won't work. Whether it should is an interesting question. Strictly speaking, disabling the
> setting should prevent the sound volume from following the voice volume, but it doesn't say anything about snapping the volume back to maximum. The user
> might have had the sound volume set to something lower before they enabled follow. However, it also doesn't really make sense to me to keep track of the
> level the user set before enabling follow just so we can restore it later.

I hadn't considered it from that prospective, but I concede the point.

Why the objection to keeping track? I'm not saying I disagree, but I wonder what 
your thinking is.

If your objection is to storing a representation of the volume in nvda.ini, one 
possible way around this would be to keep track, but only for the session.

When the checkbox is checked, or NVDA starts with the box checked, record what 
the volume was before matching it to speech.

When the box is unchecked, or NVDA shuts down, reset the volume to what was 
remembered. This assumes that if NVDA is shut down with the box checked, but the 
setting is not in the saved config, the system volume mixer will remember the 
application's volumes on the next load.

Of course, that doesn't take into account unusual shutdowns.

**Comment by jcsteh** [comment]
Yeah, I'm slightly less bothered by storing it in the session, but the UX isn't great, since then the behaviour is inconsistent depending on whether the user enabled and disabled in the same session or not.

The other challenge is how to apply the volume change when disabling. One way is to create a temporary WavePlayer using the sounds session just to adjust the session volume. We'd need to do this when applying in the dialog, when switching config profiles and when reverting the config. Perhaps a simpler way I'm considering is just to track the last state in the WavePlayer class.

File: source/gui/settingsDialogs.py [hunk 3]
```diff
@@ -3032,6 +3032,18 @@ def __init__(self, parent):
...
+		#  Advanced settings panel.
+		label = _("Volume of NVDA sounds follows voice volume (requires WASAPI)")
+		self.soundVolFollowCheckBox: wx.CheckBox = audioGroup.addItem(
+			wx.CheckBox(audioBox, label=label)
+		)
+		self.bindHelpEvent("SoundVolumeFollowsVoice", self.soundVolFollowCheckBox)
+		self.soundVolFollowCheckBox.SetValue(
+			config.conf["audio"]["soundVolumeFollowsVoice"]
+		)
+		self.soundVolFollowCheckBox.defaultValue = self._getDefaultValue(
```
Comment (seanbudd) [inline source/gui/settingsDialogs.py:3045]: we should probably disable/enable `soundVolFollowCheckBox` when `wasapiCheckBox` is toggled. 

I think this pattern exists in the Braille settings panel (e.g. `onShowCursorChange`)

File: source/gui/settingsDialogs.py [hunk 4]
```diff
@@ -3129,6 +3144,9 @@ def onOpenScratchpadDir(self,evt):
 		path=config.getScratchpadDir(ensureExists=True)
 		os.startfile(path)
 
+	def onWasapiChange(self, evt):
```
Comment (seanbudd) [inline source/gui/settingsDialogs.py:3147]:
```suggestion
	def onWasapiChange(self, evt: wx.SpinEvent):
```

File: source/gui/settingsDialogs.py [hunk 5]
```diff
@@ -3129,6 +3144,9 @@ def onOpenScratchpadDir(self,evt):
 		path=config.getScratchpadDir(ensureExists=True)
 		os.startfile(path)
 
+	def onWasapiChange(self, evt):
+		self.soundVolFollowCheckBox.Enable(evt.IsChecked())
```
Comment (seanbudd) [inline source/gui/settingsDialogs.py:3148]: should this also be set to false if wasapi is disabled?

```suggestion
		wasapiEnabled = evt.IsChecked()
		self.soundVolFollowCheckBox.Enable(wasapiEnabled)
		if not wasapiEnabled:
			self.soundVolFollowCheckBox.SetValue(False)
```

**Review by seanbudd — COMMENTED** [review]
Looks almost ready, thanks @jcsteh

File: source/gui/settingsDialogs.py (hunk 5 above)
Comment (jcsteh) [reply source/gui/settingsDialogs.py:3148]: I'm not convinced we should do this. This means that even if the user previously enabled it, we'll lose that setting when they disable and re-enable WASAPI. If they're just trying to test something without WASAPI for some reason, I don't see why they should have this other setting reverted to default when they re-enable WASAPI.

That said, if you feel really strongly about this, I'll relent and make the change. :)

File: source/gui/settingsDialogs.py (hunk 4 above)
Comment (jcsteh) [reply source/gui/settingsDialogs.py]: I'm fairly sure checkbox events are CommandEvent, not SpinEvent, so I went with CommandEvent.

**Review by seanbudd — APPROVED** [review]
Thanks @jcsteh

**Review by Qchristensen — APPROVED** [review]
Docs changes read well, good work!

**Comment by brunoprietog** [comment]
Hi, I am trying this, but I feel that the volume of the tones does not go down linearly. With the option disabled, if I lower the volume of NVDA Sounds and NVDA to the same level using the Windows sound mixer, the sound is heard at a similar intensity. On the other hand, if I enable the option and lower the volume of the NVDA voice, the sounds remain louder than the voice at all times, except when the voice volume is 100.

**Comment by jcsteh** [comment]
What voice are you using? This is going to depend on the voice somewhat, since the voice volume is controlled by the voice, not NVDA or Windows. If the voice volume isn't entirely linear, then there is going to be a mismatch. It seems to be fairly linear with eSpeak.

**Comment by brunoprietog** [comment]
Yes, I think that is the cause. I just tried Espeak and it works fine. Is there a way to make the NVDA sound unified? That is, if I lower the volume of NVDA in the Windows sound mixer, it lowers the volume of the voice and the volume of the tones equally, just as it works when you don't have Wasapi enabled. Thanks!

**Comment by jcsteh** [comment]
There isn't a way to do that, no. They are separate sessions so that users can adjust them separately, but the flip side of that is that they are... well... adjustable separately.

**Comment by brunoprietog** [comment]
Sure, I understand. But isn't there a way I can be combined in one session? Just like it works in previous versions. That allows you to reduce the volume of NVDA from the Windows sound mixer, without having to worry about lowering the volume of NVDA and NVDA sounds. I understand that's the point of this PR, but as I was telling you, it doesn't work well with other synthesizers. Thanks!

**Comment by jcsteh** [comment]
The point of this PR was that users wanted a way to do this which used the voice volume, since the voice volume is easier to control than the Windows sound mixer volume; e.g. you can use the synth settings ring. I don't really see a way we can support all of these different things at once when some of them are in conflict. I guess a setting might be possible, but these settings would have to be mutually exclusive, which is confusing at best.

What synth are you using? I wonder whether the correct solution is to fix the synth driver to set the voice volume in a more useful way.

**Comment by brunoprietog** [comment]
Shure. I'm using the [IBM TTS synthesizer](https://github.com/davidacm/NVDA-IBMTTS-Driver).

**Comment by jcsteh** [comment]
Okay. As a third party synth, we have no control over that.

**Comment by brunoprietog** [comment]
Exactly, I completely understand. That's why I was asking about a possibility to have the tones and the voice in the same session, so as not to have this kind of problems. Instead of having each session separately

**Comment by jcsteh** [comment]
Which takes us back to people not being able to control them separately.

**Pull request merged**